DigitalLearningSolutions.Data/Services/DelegateApprovalsService.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace DigitalLearningSolutions.Data.Services
+{
+    using DigitalLearningSolutions.Data.DataServices;
+    using DigitalLearningSolutions.Data.Models.CustomPrompts;
+    using DigitalLearningSolutions.Data.Models.User;
+
+    public interface IDelegateApprovalsService
+    {
+        public List<(DelegateUser, List<CustomPromptWithAnswer>)>
+            GetUnapprovedDelegatesWithCustomPromptAnswersForCentre(int centreId);
+    }
+
+    public class DelegateApprovalsService : IDelegateApprovalsService
+    {
+        private readonly IUserDataService userDataService;
+        private readonly ICustomPromptsService customPromptsService;
+
+        public DelegateApprovalsService(IUserDataService userDataService, ICustomPromptsService customPromptsService)
+        {
+            this.userDataService = userDataService;
+            this.customPromptsService = customPromptsService;
+        }
+
+        public List<(DelegateUser, List<CustomPromptWithAnswer>)> GetUnapprovedDelegatesWithCustomPromptAnswersForCentre(int centreId)
+        {
+            var users = userDataService.GetUnapprovedDelegateUsersByCentreId(centreId);
+            var usersWithPrompts =
+                customPromptsService.GetCentreCustomPromptsWithAnswersByCentreIdForDelegateUsers(centreId, users);

```

Comment:
Would it be feasible to pull the logic out of this GetCentreCustomPromptsWithAnswersByCentreIdForDelegateUsers method and put in into this service? It feels like the actual logic performed in that method is just the select statement on the passed in delegateUsers, which could be done here instead. If it makes more sense to keep it as is, then feel free to leave it.

Reply:
The select statement uses a private method from the CustomPromptsService, which I'm inclined to leave as private so this method mirrors the GetCentreCustomPromptsWithAnswersByCentreIdAndDelegateUser method.

Reply:
It's actually really easy to change the names, just change the interface on the DelegateApprovalService to:
```
public List<(DelegateUser delegateUser, List<CustomPromptWithAnswer> prompts)>
            GetUnapprovedDelegatesWithCustomPromptAnswersForCentre(int centreId);
```

(and potentially the method too)